### PR TITLE
feat: :sparkles: added new controllers, env,  route updates, refactor connection strem

### PR DESCRIPTION
**Pull request by Yusuf-R**
​
## Description

StudentController for sign-in, activateProfile and login
Authentication Controller for Basic Authentication using tokens
Implemented connection stream for Redis and MongoDB client using .env file
​
## Related Issue

​
## Motivation and Context

​
## How Has This Been Tested?

​
## Screenshots (if appropriate - Postman, etc):
​
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
​
## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/routes/index.js
Comment (Dev-wonderful) [inline src/routes/index.js:1]: I think you should move the student router to a different file then import it into this index file with the name `studentRoute` then use `router.use()` to point to the `studentRoute` handler, almost like blueprint in flask

In the `routes/student.js` file
```
const express = require('express');
const studentRouter = express.Router()

studentRouter.post('/signin', studentController.activateProfile);

and so on
```

in the `routes/index.js` file
```
const { studentRouter } = require('path/to/routes/student.js');
const express = require('express');
const router = express.Router();

router.use('/api/v1/studentportal', studenRouter);

and so on
```

You can check for the credibility of this approach, this way this file won't get too cluttered

Comment (Yusuf-R) [reply src/routes/index.js:1]: Okay, that's nice, will refactor

File: src/controllers/StudentsController.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    }
+    if (!email) {
+      return res.status(400).json({ error: 'Missing email' });
+    }
+    const userData = {
+      firstName,
+      email,
+    };
+    try {
+      await dbClient.isAlive();
```
Comment (adobki) [inline src/controllers/StudentsController.js]: As discussed, you need to test/use the return value of this expression, as nothing after this line would work if it is falsey. Remember that the logic we implemented in the `isAlive` method actually catches/handles all its errors, so your try/catch statement here would never actually have any errors to catch from same. A false is simply returned instead.

File: src/controllers/StudentsController.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      const userData = {
+        matricNo,
+        password,
+      };
+      await dbClient.isAlive();
+      const user = await Student.findOne({ matricNo });
+      if (!user) {
+        return res.status(400).json({ error: 'User not found' });
+      }
+      if (user.status !== statuses[1]) {
```
Comment (adobki) [inline src/controllers/StudentsController.js:178]: This line should be `if (user.status === statuses[0]) {`, as verification is for status[0] only, right?

File: src/routes/index.js
Comment (adobki) [reply src/routes/index.js:1]: > . . .will refactor

Waiting for this before approving the PR. . . @Yusuf-R

File: src/routes/index.js [hunk 3]
```diff
@@ -7,6 +6,7 @@ const express = require('express');
 const router = express.Router();
 // import the user controller
 const userController = require('../controllers/UsersController');
```
Comment (adobki) [inline src/routes/index.js]: Remember to delete this and the referenced controller.

File: src/controllers/StudentsController.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+/* eslint-disable no-undef */
+/* eslint-disable no-unused-vars */
+/* eslint-disable consistent-return */
+/* eslint-disable import/no-extraneous-dependencies */
+// bcrypt for password hashing
+const bcrypt = require('bcrypt');
+// import the user model
+const { Student } = require('../models/student');
+const { enums } = require('../models/base');
+const dbClient = require('../utils/db');
```
Comment (adobki) [inline src/controllers/StudentsController.js:12]: I think you should import `dbClient` from the `base` model, as opposed to directly from the `db` util.

File: src/controllers/StudentsController.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      return res.status(400).json({ error: 'Missing token' });
+    }
+    if (!email) {
+      return res.status(400).json({ error: 'Missing email' });
+    }
+    if (!password) {
+      return res.status(400).json({ error: 'Missing password' });
+    }
+    try {
+      await dbClient.isAlive();
```
Comment (adobki) [inline src/controllers/StudentsController.js]: Same as above, but in this case, you should probably just take out the try/catch altogether and simply use an if statement.

File: src/controllers/StudentsController.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    }
+    if (!password) {
+      return res.status(400).json({ error: 'Missing password' });
+    }
+    try {
+      const userData = {
+        matricNo,
+        password,
+      };
+      await dbClient.isAlive();
```
Comment (adobki) [inline src/controllers/StudentsController.js]: Same as above, . . .use an if statement. . .

File: src/controllers/AuthController.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+// token key expiration 24hrs
+const EXP = 60 * 60 * 24;
+
+class AuthController {
+  // create token base of user object credentials and store it in redis
+  static async createXToken(userID) {
+    const token = uuidv4();
+    const key = `auth_${token}`;
+    try {
+      await redisClient.set(key, userID, EXP);
```
Comment (adobki) [inline src/controllers/AuthController.js:56]: All `redisClient` calls should be preceeded by an `isAlive` call in the same block/scope, yeah? Just wondering. . .

**Review by adobki — CHANGES_REQUESTED** [review]
Good job, bro! All or at least some of these need to be addressed before this is approved though, so that a separate PR won't need to be created just for them. Thanks.

File: src/controllers/StudentsController.js (hunk 2 above)
Comment (Yusuf-R) [reply src/controllers/StudentsController.js:178]: At this stage, it is login, we believe the user status should now be active and not at the init stage, thus preventing a user at init stage from login in directly

File: src/controllers/AuthController.js (hunk 7 above)
Comment (Yusuf-R) [reply src/controllers/AuthController.js:56]: I implemented the isAlive check for set, get and del in the methods for the Redis class. 
thus not needing a preceding check

File: src/controllers/StudentsController.js (hunk 1 above)
Comment (Yusuf-R) [reply src/controllers/StudentsController.js]: code updated

File: src/routes/index.js (hunk 3 above)
Comment (Yusuf-R) [reply src/routes/index.js]: folder structure updated

File: src/controllers/StudentsController.js (hunk 4 above)
Comment (Yusuf-R) [reply src/controllers/StudentsController.js:12]: it's the utility folder

File: src/controllers/StudentsController.js (hunk 5 above)
Comment (Yusuf-R) [reply src/controllers/StudentsController.js]: updated

File: src/controllers/StudentsController.js (hunk 6 above)
Comment (Yusuf-R) [reply src/controllers/StudentsController.js]: updated

File: src/routes/index.js
Comment (Yusuf-R) [reply src/routes/index.js:1]: Done

**Review by adobki — APPROVED** [review]

**Pull request merged**